### ISSUE TITLE
currentSession() returns no current user?

### ISSUE THREAD

**Issue by vsrivatsan89** [aws/aws-amplify#399]
I'm trying to test with the following code in a component that is wrapped by the withAuthenticator() HOC. Federated login works. Once logged in, I get the "No current user" error when running the following code, but the user object is getting populated. 

Am I not understanding how the library works? Shouldn't the currentSession return JWT, acesstoken, id?

Ultimately, I want to be able to get email from Facebook as identity provider. 

```
componentDidMount = () => {
    window.LOG_LEVEL = 'DEBUG'
    console.log("About to mount")
    Auth.currentSession().then(function(session) {
      console.log(JSON.stringify(session))
    }, function(err) {
      console.log(err)
    })
    Auth.currentAuthenticatedUser().then(function(user) {
      console.log("USER: " + JSON.stringify(user))
    })
  }
```

Output:
```
No current user
USER: {"Name":"x x"}
```

**Comment by jhorman** [comment]
Getting the same thing. `currentSession` works with normal login, but with federated login I get `No current user`

**Comment by powerful23** [comment]
@vsrivatsan89 The ```currentSession``` only works when you logged in with Cognito Userpool. We will work on the naming to make it more clear. Thanks!.

**Comment by epicfaace** [comment]
@powerful23 is there an alternative for `currentSession` when you are using federated identities?

**Comment by powerful23** [comment]
@epicfaace you can retrieve the current user by ```currentAuthenticatedUser``` for federated identities and ```currentCredentials``` to get the credentials

**Comment by dev-jpnobrega** [comment]
@powerful23 currentAuthenticatedUser returns "not authenticated" for federated users

**Comment by dev-jpnobrega** [comment]
aws-amplify  0.2.14
aws-amplify-react 0.1.38

After the federated authentication it returns the access_token, however Auth.currentAuthenticatedUser () returns "not authenticated"

**Comment by powerful23** [comment]
@dev-jpnobrega can you turn on the debug mode by ```window.LOG_LEVEL='DEBUG'```? And also could you check your localStorage to see if there is an item named like ```aws-amplifyfederatedUser```

**Comment by dev-jpnobrega** [comment]
Hi @powerful23, 


![image](https://user-images.githubusercontent.com/28389807/38563563-8c64605c-3cb3-11e8-9c61-c5eda359bf50.png)

![image](https://user-images.githubusercontent.com/28389807/38563594-a043c5b8-3cb3-11e8-8f08-96e59b0a96b6.png)


[localhost-1523370483986.log](https://github.com/aws/aws-amplify/files/1895063/localhost-1523370483986.log)

**Comment by dev-jpnobrega** [comment]
If I use the Auth.federatedSignIn it creates the localstorage aws-amplify-cachefederatedInfo and aws-amplify-cacheCurSize, but Auth.currentAuthenticatedUser returns not authenticated

aws-amplifyfederatedUser is not created

**Comment by powerful23** [comment]
@dev-jpnobrega do you have cognito identity pool and pass its id into Amplify? You are using a federated provider named "NaturaAD" but did you set it as a provider in that cognito identity pool? If you didn't, you cannot get a valid credential from Cognito.

**Comment by mlabieniec** [comment]
Closing this due to inactivity, please let us know if you are still experiencing the issue.

**Comment by epicfaace** [comment]
I am still experiencing this issue. When I keep the window open for a while, then it appears that my session times out and calling currentSession() returns no active user.

**Comment by powerful23** [comment]
@kanteankit you can get the jwt token by using the `Cache`:
```js
import { Cache } from 'aws-amplify';

Cache.getItem('federatedInfo').then(federatedInfo => {
     const { token } = federatedInfo;
});
```

Sorry we didn't mention this in the doc, will send a pr for that.

**Comment by powerful23** [comment]
@kanteankit how are you using that jwt token to make calls to API. As I know the jwt tokens are only used to fetch a AWS credentials from your federated identity pool.

**Comment by elorzafe** [comment]
Hi @kanteankit how are you configuring you Authorizers on API GW?

**Comment by kanteankit** [comment]
@elorzafe 
As mentioned in my previous comments,I am familiar with Amazon Cognito and  I am already using `Auth.signIn` for authenticating via email. That email authentication is creating a user in Cognito User Pool and giving me a JWT token that I can use in Authorization header.

The entire point of my question is how can I do a similar thing with `Auth.federatedSignIn`. After calling `Auth.federatedSignIn`, what should I do to get an Authorization token?

Note: I cannot use awsmobile-cli.

**Comment by elorzafe** [comment]
@kanteankit Your should get an Authorization token from your auth provider. Once you get that token you can do this.
```
var provider_url = 'my.auth-provider.com';
await Auth.federate(provider_url, { token: id_token });
var credentials = await Auth.currentUserCredentials();
```

**Comment by dnalbach** [comment]
@elorzafe Auth does not have a 'federate' method? I checked the source code for Auth in Amplify. There is only a federatedSignIn method, which @kanteankit has already done. Can you elaborate?

**Comment by akashbits** [comment]
@elorzafe any update ?

**Comment by mlabieniec** [comment]
@dnalbach the method is [federatedSignin](https://aws.github.io/aws-amplify/media/authentication_guide#enabling-federated-identities) apologies for the confusion. The docs here explain what to do with the token, you send it to them method as an arg along with the provider:

```js
return Auth.federatedSignIn(
        // Initiate federated sign-in with Google identity provider 
        'google',
        { 
            // the JWT token
            token: id_token, 
            // the expiration time
            expires_at 
        },
        // a user object with profile info from the provider (see docs)
        user
    ).then(() => {
        // ...
    });
```

**Comment by Jp3rd** [comment]
@mlabieniec I am experiencing the same issue. You are still not getting what @kanteankit is saying. 

We are talking about the **JWTOKEN** from Cognito. Not the one from the identity provider(eg. Facebook, Google.) 

After using Auth.federatedSignIn it only returns the Cognito credentials Obj. So the question is, How do we get the **JWTTOKEN from Cognito** after the use of Auth.federatedSignIn?

**Comment by powerful23** [comment]
@Jp3rd sorry for the confusion but with `Auth.federatedSignIn` you are only getting the credentials but not a Cognito User back so there is no JWT token from Cognito but only tokens from your federation providers. 

https://github.com/aws/aws-amplify/wiki/FAQ#how-can-i-get-jwt-token-when-using-amplify-to-get-federated-users-login

**Comment by eyalc4** [comment]
@kanteankit 
have you manage to make it work and get a token to call AWS services after doing, Auth.federatedSignIn ?

**Comment by Nickadiemus** [comment]
@eyalc4 I can with my current solution I'm using.

**Comment by akash87** [comment]
@Nickadiemus can you provide details as to how do you get access token for federated logins ?

**Comment by akash87** [comment]
Thanks @nickadiemus, I thought you made it work without hosted UI, that's the main issue here. Looks like there is no way of getting AWS token without using hosted UI.

**Comment by Nickadiemus** [comment]
@akash87 Oh yeah sorry! That's the only work around I've gotten to work 🤕

**Comment by flaviocordova** [comment]
I'm facing the exact same problem and, I have to say, I feel a little frustrated...

I expected signIn and federatedSignIn to to be only different ways to authenticate the user... I understand user and identity pools have different goals but at this point (as a developer), I just need a simple, unified way to authenticate users, be able to identify them (from the token, server-side) and use services.

PS: @mlabieniec, this issue was closed by inactivity but I think it should be reopened...

**Comment by Nickadiemus** [comment]
@flaviocordova I agree do agree.

**Comment by alexzaganelli** [comment]
I'm struggling with it, no way without hosted UI... I'm very frustrated!!

**Comment by powerful23** [comment]
Hey guys we had a meeting with Cognito team and found out there is a way to use your own UI to sign in with the federated providers and create an user in your Cognito User Pool:  https://github.com/aws-amplify/amplify-js/issues/1316

**Comment by giaendong** [comment]
facing the same problem but after changed the amplify configuration from
`Amplify.configure({
  Auth: {
    region: 'xx-xxxxx',
    identityPoolId: 'xx-xxxxxx',
    userPoolId: 'xx-xxxx',
    userPoolWebClientId: 'xxxxxxxxx',
    mandatorySignIn: false,
    storage: xxxxx,
  }
});`
to
`Amplify.configure({
  Auth: {
    region: 'xx-xxxxx',
    userPoolId: 'xx-xxxx',
    userPoolWebClientId: 'xxxxxxxxx',
    aws_appsync_authenticationType: 'AMAZON_COGNITO_USER_POOLS'
  }
});`
currentSession() now worked

**Comment by byio** [comment]
hey guys, i'm using the Auth.signIn() method which returns a CognitoUser object and among its properties, the signInUserSession has a value of null.

and when i ran `await Auth.currentSession()` i got 'No current user' as well.

if it helps, i'm following a tutorial and this is the [step](https://serverless-stack.com/chapters/load-the-state-from-the-session.html) i'm currently on when i encountered this issue.

**Comment by SajidNawaz993** [comment]
Hi .. i have receive jwt token in case of fb login in react native  but i have no idea what to do next . frist time user get from social login in aws but second time 403 error no current user . Please guide me . Thanks

**Comment by rAnAaRsl** [comment]
@SajidNawaz993 did u find any solution ?

**Comment by mayteio** [comment]
Far out, I can't believe this is still open with no real resolution. We have got up until `federatedSignIn` and now need JWT to access the Amplify GraphQL API. Have tried this with both Facebook login and a developer identity.

Can someone confirm from AWS that there is no other solution than to create a hosted UI?

**Comment by powerful23** [comment]
@mayteio I'm sorry about the bad experience when using `Auth.currentSession` with federated sign in. We are currently working on to provide a standard way to get the session, user, credentials for both Cognito users and federated users. That should resolve the issue to get the JWT token when users logged in via `federatedSignIn`. 

For now as a workaround you can directly get the federated token from the `localStorage`(web)/`AsyncStorage`(RN):
```js
const token = JSON.parse(storage.getItem('aws-amplify-federatedInfo')).token;
```

**Comment by eyalc4** [comment]
dont take it as 100% bullet proof, as i am **not** an AWS employee, this summary is only according to my experience with AWS cognito.
**TL:DR**
1. AWS Cognito documentation is bad, it is confusing regarding its components (user pool, identity pool), and miss-leading about what it can and can't do and how. 
2. The cognito SDK, isn't sufficient, isn't well enough documents, and rarely have code examples - reference guide, isn't enough.
3. Its high cost, isn't its monthly usage, it’s the too much time, took too many developers, to figure out it doesn't answer their basic needs. 
4. If you are looking for authentication solution, and want to write your own UI - it is probably isn't the solution for you – currently, hope they will chose to support it in future.

**Cognito overview:**
1. User pools – create users authentication records:
    a.	Save user attributes
    b.	Authenticate users with user and password
    c.	Email / phone verification
    d.	Access the user details from SDK
2.	Federation identity (identity pools)
    a.	Required in order to get temporary token, to access amazon resources (S3, Api gateway etc)
    b.	It federate – i.e. gives temporary token, to users from external sources:
       i.	User pools
       ii.	Facebook App / Google App
    c.	Identity record being created – in the (federated) identify pool, NOT in users pool, but you cannot 
         get user details by token, credentials, or any other identity field.
    3.	Cognito Buildin-UI enable you to have a flow of creating user in User pool, and get federation identity created when needed to access AWS services. The user can be created with both user & password or with Social services (fb, google etc)
This UI, is the ONLY method, to create a user in the user pool, from social login.
    4.	If you want to implement your own UI, you depends on the SDK, and can create user only with user & password, not through social login.
    5.	The SDK is limited, and different from one language to another:
      a.	you cannot create a user from social login 
      b.	you cannot create a user without a password
      c.	you cannot fetch user information from Federated identity
      d.	you can fetch user information from user pool, but this user exist only if you user user & password / you used the build-in UI to create it.

**Recommendation:**
    1.	If you are building a web-app, NOT a mobile app, and doesn’t need advanced user registration / login, the Built-in UI may be good enough for you, and you may benefit from Cognito full solution, having all your users in the user pool, and get the access from federated.
    2.	If you don’t need to have your own users, and just want to enable social users to get temporary access to AWS resources – you just need federated login.
    3.	If you don’t want to use the Build-in UI, and want a social login – you will not be able to use the cognito SDK to manage your users, and you will need to have your own solution for one unified users table + basic abilities, example of problematic use cases: 
    a.	user registered with email + password, and later on, sign-in with facebook connect, you need to know it is the same user. 
    b.	You want to have user id for each user.
    c.	You have a server side, that needs to know who is the currently logged in user, and you don’t want to make it integrate with multiple social services, but with a single users API.

**Comment by powerful23** [comment]
@mayteio yes, so the `federatedSignIn` method only works with Cognito Federation Identity Pool to by passing the JWT token from third provider to get a temporary AWS credentials. If you want to do federation with Cognito User Pool and get a JWT token from it, you have to use Cognito Hosted UI.

@eyalc4 thanks for your feedback.

**Comment by nayanbhana** [comment]
Please update your documents to reflect this AWS. Very frustrating.

**Comment by mayteio** [comment]
Thanks @eyalc4 for that awesome experiential deep dive. 

Would a solution be to;

1. on 3rd party login
2. check for username
3. if no username, create a user in user pool
4. email password to user

and do this in a Lambda? Or is that not possible?

If they log in again via 3rd party, only give them access again if they successfully get an authcode back from the oAuth flow?

**Comment by revmischa** [comment]
I'm using cognito hosted UI and trying to authenticate to appsync. I followed these instructions: https://aws-amplify.github.io/docs/js/react#add-graphql-backend
With the same configuration, the only difference is that I have federated auth. 
I also get a "No current user" error, even though the hosted cognito UI says I'm logged in (via google).
As a developer I'd expect once someone signs in via the cognito hosted UI, I'd be able to make appsync requests for my appsync service configured to permit via cognito. The fact someone is signed in via federated identity should not matter.

**Comment by powerful23** [comment]
@revmischa how are you using the Cognito Hosted UI? Did you follow this doc to configure it in Auth? https://aws-amplify.github.io/docs/js/authentication#using-amazon-cognito-hosted-ui

**Comment by GenghisJuan** [comment]
I too have been hacking around here trying to find out a solution. Just to be clear, if I am using react-native + amplify, there is no way to have a user sign up with Facebook and have that user in my Cognito user pools? This is like the most basic use case, hard to believe it isn't supported.

**Comment by eyalc4** [comment]
@GenghisJuan you right, it is the most basic - except if u develop a website, and can use the hosted UI.
I was also sure that it had to be possible - and if I didn’t got it to work, I must have done something wrong, it took me 5 weeks to give up and share my lesson here in details to save other developers the frustration.  (Scroll up if u missed it)
;) so feel free to waste your time, don’t forget to let us know how was it...

**Comment by alexvuta** [comment]
I am facing the same issue. 
I use Facebook as provider for login, but i can't get a session after calling federatedSignIn.
There is not a way to get a user session (idToken, RefreshToken..).

**Comment by abiodunjames** [comment]
I'm facing this same problem using `Auth.federatedSigin` with google sign in. `Auth.currentCredentials` returns an object that can not be used to used to access resources on AWS

**Comment by powerful23** [comment]
@abiodunjames can you provide more details? Did you have the right IAM policy for your auth role in the Cognito Federated Identity Pool? You can also turn on the debug mode by `window.LOG_LEVEL='DEBUG'`

**Comment by donaldev** [comment]
I'm experiencing the same issue. To clear this up when you create an API and add an Authenticator to it, it expects a Token to prove that the user accessing the API came from cognito, i.e. it is safe to provide access. 

When you call Auth.currentSession() after a cognito user pool login you are given the following response
`CognitoUserSession {idToken: CognitoIdToken, refreshToken: CognitoRefreshToken, accessToken: CognitoAccessToken, clockDrift: 0}
accessToken: CognitoAccessToken {jwtToken: "xxxxxxx", payload: {…}}
clockDrift: 0
idToken: CognitoIdToken
jwtToken: ****[THIS IS USED FOR API CALLS]****
payload: xxxx
__proto__: CognitoJwtToken
refreshToken: CognitoRefreshToken {token: "xxxxxxxx"}
__proto__: Object`

The idToken->jwtToken is the token needed to authorize API calls.

When Auth.currentSession() is called after a federated sign in, the following is returned : 
`no current user`.

How can we get a valid jwt token from a federated sign-in user?

Has there been any updates?

**Comment by ahsanwtc** [comment]
having the same problem

**Comment by abiodunjames** [comment]
This is what I eventually did and I was able to access aws resources using Google federated identity
```js

 const profile = googleUser.getBasicProfile(); // etc etc

      let user = {
        name: profile.getName(),
        email: profile.getEmail()
      };
      let auth = googleUser.getAuthResponse();
      const {
        expres_in,
        idpId,
        id_token // Optional, the identity id specified by the provider
      } = auth;
      Auth.federatedSignIn(
        idpId,
        {
          id_token,
          expires_at: 108000 + new Date().getTime()
        },
        user
      )
        .then(cred => {
          return Auth.currentAuthenticatedUser();
        })
        .then(user => {
       console.log(user)
        })
        .catch(e => {
          console.log(e);
        });
```

**Comment by notsoluckycharm** [comment]
@abiodunjames Can you elaborate? Are you calling aws services directly with v4 signing? This still shouldn't work to create an Authorization header capable of being used with Cognito autherizers between ApiGateway and Lambda.

**Comment by DannyHinshaw** [comment]
Hopefully this will help some other poor souls. We are using cognito pools I had to change my `awsconfiguration.json`. When amplify generated it it looked like this:

```json
{
    "UserAgent": "aws-amplify-cli/0.1.0",
    "Version": "1.0",
    "IdentityManager": {
        "Default": {}
    },
    "AppSync": {
        "Default": {
        "ApiUrl": "https://nuiqzh65tfjksd4lajfie5ff2i.appsync-api.us-east-1.amazonaws.com/graphql",
        "Region": "us-east-1",
        "AuthMode": "AMAZON_COGNITO_USER_POOLS",
        "ApiKey": "da2-75qmsylif5ew5h5zyrcl6p4hdy"
        }
    }
} 
```

I needed to manually add `CredentialsProvider` and `CognitoUserPool ` so the new config looks like so:

```json
{
  "UserAgent": "aws-amplify-cli/0.1.0",
  "Version": "1.0",
  "IdentityManager": {
    "Default": {}
  },
  "AppSync": {
    "Default": {
        "ApiUrl": "https://nuiqzh65tfjksd4lajfie5ff2i.appsync-api.us-east-1.amazonaws.com/graphql",
        "Region": "us-east-1",
        "AuthMode": "AMAZON_COGNITO_USER_POOLS",
        "ApiKey": "da2-75qmsylif5ew5h5zyrcl6p4hdy"
    }
  },
  "CredentialsProvider": {
    "CognitoIdentity": {
      "Default": {
        "PoolId": "us-east-1:8e535f8d-9a88-556a-88dd-1a23da7dd987",
        "Region": "us-east-1"
      }
    }
  },
  "CognitoUserPool": {
    "Default": {
      "PoolId": "us-east-1_gsdsBZ0zk",
      "AppClientId": "1f5kj2l3fhjkl2u8f7i2ijfdskl2hsnt89r62",
      "AppClientSecret": "s2kieg3afli2ufh2fo8su45hf4ka5naqvcd85559j5",
      "Region": "us-east-1"
    }
  }
}
```

(all data has been obfuscated)

**Comment by okarlsson** [comment]
I got the same error while using the Authenticator from aws-amplify-react, after logging in with facebook `Auth.currentSession()` returned `No current user`.

But if I use `Auth.federatedSignIn()`, `Auth.currentSession()` returns an access token that I can use. And I can skip displaying the hosted UI by naming a provider as a parameter `Auth.federatedSignIn({provider: 'Facebook'})`

So do **not** use the facebook sign in button on the  `<Authenticator />` component if you want an access token. **Do** use the `Auth.federatedSign({provider: 'provider-name'})` to get an access token and skip the UI.

Here are some sample code, replace the variables for your user pool with your own:

```
import React from "react";
import Amplify, { Auth, Hub } from "aws-amplify";

Amplify.configure({
  Auth: {
    region: "eu-central-1",
    userPoolId: "eu-central-1_xxxxx",
    userPoolWebClientId: "xxxxxxxxx"
  }
});

const oauth = {
  domain: "your_domain.auth.eu-central-1.amazoncognito.com",
  scope: ["email", "profile"],
  redirectSignIn: "http://localhost:3000",
  redirectSignOut: "http://localhost:3000",
  responseType: "code"
};

Auth.configure({ oauth });

class App extends React.Component {
  state = {
    session: null
  };

  constructor() {
    super();
    Hub.listen("auth", ({ payload: { event, data } }) => {
      switch (event) {
        case "signIn":
          console.log("Sign in");
          this.setSession();
          break;
        case "signOut":
          console.log("Sign out");
          break;
        default:
          break;
      }
    });
  }

  setSession = async () => {
    const session = await Auth.currentSession();
    this.setState({ session });
  };

  render() {
    const { session } = this.state;

    return (
      <div className="App">
        <button onClick={() => Auth.federatedSignIn({ provider: "Facebook" })}>
          Facebook
        </button>
        <button onClick={() => Auth.federatedSignIn()}>Open Hosted UI</
        <p>{JSON.stringify(session)}</p>
      </div>
    );
  }
}

export default App;

```

**Comment by scaralfred** [comment]
Sad to see that this issue was not addressed yet. Still no way to use graphQl in a unified way while using federated.signIn.

**Comment by moseley** [comment]
When calling `Auth.currentSession()` I was receiving a `No current user` response. In my case the reason this was happening was that I created a new user from the console and even though I check the box to "Mark email as verified?" the account status was `FORCE_CHANGE_PASSWORD`. A verified user will have a status of `CONFIRMED`. You can do this with a test user in AWS CLI
```aws cognito-idp admin-confirm-sign-up \
   --region YOUR_COGNITO_REGION \
   --user-pool-id YOUR_COGNITO_USER_POOL_ID \
   --username YOUR_USER_EMAIL
```

**Comment by diegodurli** [comment]
Can we please reopen this issue? It's still definitely relevant given comments in this thread.

**Comment by bneigher** [comment]
Why is this still a bug?  This is a pretty severe issue for all 3rd party auths...

**Comment by bijanmmarkes** [comment]
Any updates on this issue?

**Comment by adamcasey** [comment]
> When calling `Auth.currentSession()` I was receiving a `No current user` response. In my case the reason this was happening was that I created a new user from the console and even though I check the box to "Mark email as verified?" the account status was `FORCE_CHANGE_PASSWORD`. A verified user will have a status of `CONFIRMED`. You can do this with a test user in AWS CLI
> 
> ```
> aws cognito-idp admin-confirm-sign-up \
>    --region YOUR_COGNITO_REGION \
>    --user-pool-id YOUR_COGNITO_USER_POOL_ID \
>    --username YOUR_USER_EMAIL
> ```

That's exactly what was happening in my case. I was able to get `Auth.currentSession()` to actually return a user that was signed in by deleting the user and re-registering them to get a new email verification link. Once that email was confirmed, `Auth.currentSession()` gave me the current user as expected.

**Comment by Graig123git** [comment]
Still no way to use graphQl resource while using federated.signIn, I keep getting no current user, FYI am doing this in react native and i am not using hosted UI.

**Comment by simoazzo** [comment]
Is there any updates on this, please? Tahnks

**Comment by michalmikolajczyk** [comment]
Why is this issue closed? We are facing the same problem right now – can't use graphQL resource with federated signin

**Comment by easiruwa** [comment]
I'm also experiencing issues with this in a production enviroment. Are there any updates?

**Comment by marianolc** [comment]
Facing the same issue:

I've created my own identity pool in cognito. For the server/client auth workflow I followed this documentation:

https://docs.aws.amazon.com/cognito/latest/developerguide/developer-authenticated-identities.html

https://docs.amplify.aws/lib/auth/advanced/q/platform/js#identity-pool-federation

Server side code to create the token/identityId:

  var params = {
    IdentityPoolId: 'us-east-1:MY_ID_POOL',
    IdentityId: null,
    Logins: {
      'mycomp.es': jwttoken
    },
    TokenDuration: '86400'
  };

  var cognitoidentity = new AWS.CognitoIdentity();
  const logres = cognitoidentity.getOpenIdTokenForDeveloperIdentity(params, function (err, data) {
    if (err) console.log(err, err.stack); // an error occurred
    else console.log(data);           // successful response
  });

The result:

{
  IdentityId: 'us-east-1:IDENTITY_ID',
  Token: 'eyJra.... rest of the token'
}

With the Client side code, I was able to create the credentials:

        const domain = 'cognito-identity.amazonaws.com';
        const federatedUser = {
            name: 'the_user',
            email: 'the_email@gmail.com'
        };

        Auth.federatedSignIn(
            domain,
            {
                token: tokenResponse.Token, // this is the token generated
                identity_id: tokenResponse.IdentityId, // this is the identiy Id
                expires_at: 50 * 1000 + new Date().getTime() // the expiration timestamp
            },
            federatedUser
        ).then(cred => {
            // If success, you will get the AWS credentials
            console.log(cred);
            return Auth.currentAuthenticatedUser();
        }).then(user => {
            // If success, the user object you passed in Auth.federatedSignIn
            console.log("success", user);
            fetch();
        }).catch(e => {
            console.log(e)
        });

This returns no error, and the objects created seems correct:

Credentials:

{
  "accessKeyId": "xxx",
  "secretAccessKey": "yyy",
  "sessionToken": "zzz",
  "expiration": "2020-08-22T09:36:14.000Z",
  "authenticated": true
}

User:

{
  "name": "the_user",
  "email": "the_email@gmail.com"
}

But despite that I cannot access to any AWS resources yet, like graphql. I keep getting the following error:

**No current user**

**Comment by teamzz111** [comment]
Sad to see this issue closed.

**Comment by donaldev** [comment]
Why is this issue closed? Has the problem been rectified? If not the issue should stay open.

**Comment by elorzafe** [comment]
How to get Cognito Tokens when Authenticating with Social Providers

1. You need to Setup your Auth Provider with Cognito User Pools. More info [here](https://docs.amplify.aws/lib/auth/social/q/platform/js#setup-your-auth-provider)
2. On your App you have to call `Auth.federatedSignIn` in this way
    1. `Auth.federatedSignIn()` this will open Cognito Hosted UI Web App.
    2. In case you want to skip Cognito Hosted UI you can invoke from your App
        1. `Auth.federatedSignIn({provider: 'Facebook'})`
        2. `Auth.federatedSignIn({provider: 'Google'})`
        3. `Auth.federatedSignIn({provider: 'LoginWithAmazon'})`
        4. `Auth.federatedSignIn({provider: 'SignInWithApple'})`

Then you will have a user on your userpool and Cognito tokens on your App.

**Note**
If you get a token from your Authentication Provider and then use

```javascript
Auth.federatedSignIn(
    domain,
    {
        token,
        identity_id, // Optional
        expires_at: expiresIn * 1000 + new Date().getTime() // the expiration timestamp
    },
    user
)
```

You wont be able to get a Cognito Tokens and you will **No current user** error.

**Comment by obo20** [comment]
For developers who are getting this error when testing things out with localhost

Under your aws config file (described here: https://docs.amplify.aws/lib/client-configuration/configuring-amplify-categories/q/platform/js#scoped-configuration

If you have the optional "cookieStorage" object and are working with localhost, you need to have secure set to false

**Comment by Zurdge** [comment]
Just dropping this down here after I had the same "No current user" returning from "Auth.currentSession"
The issue I was having was the first federatedSignin worked, then after signing out a second signing via federatedSignin would return "No current user".
The solution I found I don't like but it works for me. A slight delay is needed between signing in and calling currentSession()...

```
      try {
        console.log("Step 1")
        await Auth.federatedSignIn({ provider: CognitoHostedUIIdentityProvider.Facebook });
      } catch (error) { reject(error) }

      console.log("Step 2")
      setTimeout(async()=>{
        try{
          const session = await Auth.currentSession();
          let accessToken = session.getAccessToken()
          resolve(accessToken);
        }catch(error){reject(error)}
        resolve()
      },100)
```

**Comment by crvargasu** [comment]
I had the same issue, and finally I use `amazon-cognito-identity-js` instead of Auth services of Amplify.  I let you the code with 2 functions, one to signUp an user and the second to SignIn. Both works in React-native. I hope this will help someone.

You also can check the documentation [here](https://github.com/aws-amplify/amplify-js/tree/master/packages/amazon-cognito-identity-js)


```
const AmazonCognitoIdentity = require('amazon-cognito-identity-js');

const poolData = {
    UserPoolId: config.USER_POOL_ID,
    ClientId: config.APP_CLIENT_ID
};

async function signUp_aws(mail, username, password) {
    var userPool = new AmazonCognitoIdentity.CognitoUserPool(poolData);

    var attributeList = [];

    var dataEmail = {
        Name: 'email',
        Value: mail,
    };

    var dataPhoneNumber = {
        Name: 'phone_number',
        Value: '+15555555555',
    };
    var attributeEmail = new AmazonCognitoIdentity.CognitoUserAttribute(dataEmail);
    var attributePhoneNumber = new AmazonCognitoIdentity.CognitoUserAttribute(
        dataPhoneNumber
    );

    attributeList.push(attributeEmail);
    attributeList.push(attributePhoneNumber);

    userPool.signUp(username, password, attributeList, null, function(
        err,
        result
    ) {
        if (err) {
            alert(err.message || JSON.stringify(err));
            return;
        }
        var cognitoUser = result.user;
        console.log('user name is ' + cognitoUser.getUsername());
    });
}

async function login_aws(mail, pass) {
 return new Promise((resolve, reject) => {
    const userPool = new AmazonCognitoIdentity.CognitoUserPool(poolData);

    let email = mail;
    let password = pass;

    let authenticationData = {
        Username: email,
        Password: password,
    };
    let authenticationDetails = new AmazonCognitoIdentity.AuthenticationDetails(
        authenticationData
    );

    let userData = {
        Username: email,
        Pool: userPool,
    };

    let cognitoUser = new AmazonCognitoIdentity.CognitoUser(userData);
    console.log(cognitoUser)


    cognitoUser.authenticateUser(authenticationDetails, {
        onSuccess: function(result) {
            var accessToken = result.getAccessToken().getJwtToken();
            var idToken = result.idToken.jwtToken;
            console.log(idToken)
        },
    
        onFailure: function(err) {
            alert(err.message || JSON.stringify(err));
        },
    });

 })
  
}

```

**Comment by mmjp-rusty** [comment]
> How to get Cognito Tokens when Authenticating with Social Providers
> 
> 1. You need to Setup your Auth Provider with Cognito User Pools. More info [here](https://docs.amplify.aws/lib/auth/social/q/platform/js#setup-your-auth-provider)
> 2. On your App you have to call `Auth.federatedSignIn` in this way
>    
>    1. `Auth.federatedSignIn()` this will open Cognito Hosted UI Web App.
>    2. In case you want to skip Cognito Hosted UI you can invoke from your App
>       
>       1. `Auth.federatedSignIn({provider: 'Facebook'})`
>       2. `Auth.federatedSignIn({provider: 'Google'})`
>       3. `Auth.federatedSignIn({provider: 'LoginWithAmazon'})`
>       4. `Auth.federatedSignIn({provider: 'SignInWithApple'})`
> 
> Then you will have a user on your userpool and Cognito tokens on your App.
> 
> **Note**
> If you get a token from your Authentication Provider and then use
> 
> ```js
> Auth.federatedSignIn(
>     domain,
>     {
>         token,
>         identity_id, // Optional
>         expires_at: expiresIn * 1000 + new Date().getTime() // the expiration timestamp
>     },
>     user
> )
> ```
> 
> You wont be able to get a Cognito Tokens and you will **No current user** error.

Am I understanding correctly, that we are expected to call Auth.federatedSignIn() TWO times during the course of an authorization or implicit grant flow? 

Meaning once to initiate the flow, and then again after a few page redirect /reload cycles on "Hub" event "signIn", to populate the user info?

User -> Client App: Click Sign In
Client App  -> Client App  : Auth.federatedSignIn(/* no args, or maybe specify your provider */);
Client App -> Auth Provider Login UI :  GET Federated Sign In Page
User  ->   Auth Provider Login UI : login
Auth Provider -> Client App : send redirect  with code|token
Client App -> Client App : extract code|token and resolve token if applicable
Client App -> Client App : Reload page without token in query params because its cleaner?
Client App -> Hub dispatch event "signin"

Now here , one would presume that your Auth.currentSession() would be populated by amplify. Because that is intuitive. But it seems that it is on the developer to bridge the gap between the federated sign in event, and initialization of  a session.

If I'm understanding correctly, the developer now has to call Auth.federatedSignIn({now we have user info to pass in}) again, but this time, passing in the user and token info present in the ephemeral hub event. 

If this is accurate than there are 2 critical design flaws  here.

1. Why doesn't amplify handle the initialization of the session on successful federated signin ? It seems that would be expected behavior for most integrators.

2. Why do we have 2 completely different methods, that do completely different things in completely different contexts, both named "Auth.federatedSignIn"?

Secret arg combinations that unlock hidden behavior in a function  is a bad pattern for contract design.

**Comment by mmjp-rusty** [comment]
The issue for me was either an invalid cookieStorage config or missing, incompatible or out dated dependencies.

Regarding the dependencies:


My "aws-amplify" version was old (2.x.x.), despite having been installed recently. This may be the result of following a tutorial to the letter on initial setup. I'm doing rapid prototyping of this application and hadn't done a thorough audit of all of my dependencies yet. 

My "peer" dependency of '@react-native-async-storage/async-storage'; which has react dependencies, was not installed. I've seen other posts indicating that a missing or old async-storage dependency can cause this "no current user" error. I'm not sure if this is truly required for an angular app.

NOTE: I'm building in angular, but some react dependencies may be necessary for the local storage to work properly. 


Also, I was configuring the cookie storage as follows:

```
Amplify.configure({
      Auth: {
        cookieStorage: {
          domain: <domain was here>,
          path: '/{i had a custom path here}',
          expires: 1,
          sameSite: "strict", // | "lax",
          secure: true
        },
```

...and either this was wrong or required some additional configuration to go with it.  

After updating my dependencies and resolving conflicts, I was still not able to get current session. Though I did not test this exhaustively before making my next change, which was to disable the cookie storage customization. 

After disabling the custom cookie storage, I was finally able to get currentSession() as of the Hub "signIn" event and this was still available after page refresh! ( as of the "configured" event )

I'm not 100% sure yet, but I believe the primary issue was an invalid cookie storage configuration. 

I was prompted to remove the cookie storage entry, after cracking up my "application" tab in chrome dev tools. I noted that there were plenty of tokens and user pool details in local storage, but none in cookie storage, despite what I had configured. This led me to believe that whatever I was specifying with cookie storage was irrelevant or possibly conflicting.

**Comment by saadahmadabbasi** [comment]
@kanteankit did you find any solution to the federarted sign in issue? 
I am still facing the issue : 21st June 2021.
I dont want to use hosted UI, withAuthenticator is also not working for user pools which has federated providers configured.

**Comment by saadahmadabbasi** [comment]
+ Adding @Jp3rd : Did you find any solution with federated sign in?
